### PR TITLE
Add HQQ support

### PR DESCRIPTION
**Pull request by mobicham**
Adds support for HQQ quantization without using the <a href="https://github.com/mobiusml/hqq/tree/master/hqq">hqq</a> lib.

The quantizer in `prototype/hqq/core.py` can output quantized tensors (`W_q`, `scale`, `zero`) that are compatible with `AffineQuantizedTensor` by setting `raw_output=False` (default value).

There's a simple example in `prototype/hqq/example.py`.

### Missing:
- Full model quantization and eval
- Hqq-mix variant (would a new layout + a new affine quantized tensor)
- Tests

### Note:
The dequantized output produced by `AffineQuantizedTensor` is a bit worse than that produced by the hqq-lib. You can check that by setting `raw_output=True`. The problem has to do with the midpoint used in the dequantization logic by `AffineQuantizedTensor` which produces zero-point values with very low magnitude.

### CODE REVIEW

File: torchao/dtypes/affine_quantized_tensor.py [hunk 1]
```diff
@@ -190,14 +193,23 @@ def from_float(
         preserve_zero: bool = True,
         zero_point_domain: ZeroPointDomain = ZeroPointDomain.INT,
         layout_type: LayoutType = PlainLayoutType(),
+        use_hqq: bool = False,
```
Comment (jerryzh168) [inline torchao/dtypes/affine_quantized_tensor.py:196]: maybe fine for now, but we could work on generalizing affine quantized tensor subclass I feel

File: torchao/prototype/hqq/example.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+q_tensor_hqq = AffineQuantizedTensor.from_float(
+        input_float=W,
+        mapping_type=MappingType.ASYMMETRIC,
+        block_size=[1, group_size],
+        target_dtype=torch.uint8,
+        quant_min=0,
+        quant_max=2**nbits - 1,
+        preserve_zero=False,#Important
+        zero_point_domain= ZeroPointDomain.FLOAT,
+        layout_type=PlainLayoutType(),
```
Comment (jerryzh168) [inline torchao/prototype/hqq/example.py:52]: does all these arguments also make sense for hqq? or is this list of args just partially used?

File: torchao/quantization/hqq.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    W_q_ao = W_q.view(shape)
+    return W_q_ao, scale_ao, zero_ao
+
+
+# Main HQQ Quantizer - simplified, no bitpacking.
+class HQQQuantizer:
+    optimize_weights = optimize_weights_proximal
+
+    @classmethod
+    def quantize(
```
Comment (jerryzh168) [inline torchao/quantization/hqq.py:104]: can this be defined as quant_primitive op (function) and added to that file? https://github.com/pytorch/ao/blob/main/torchao/quantization/quant_primitives.py

File: torchao/quantization/hqq.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    zero = zero.to(tensor.device)
+    del W_f, W_q, W_r, W_e
+    torch.cuda.empty_cache()
+
+    W_q = torch.round(tensor * scale + zero).clamp(min_max[0], min_max[1])
+    return W_q, scale, zero
+
+
+# Default: fast with early stopping
+optimize_weights_proximal = optimize_weights_proximal_legacy
```
Comment (jerryzh168) [inline torchao/quantization/hqq.py:80]: is we have different versions of this, we could define different ops/functions I feel

File: torchao/prototype/hqq/core.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+optimize_weights_proximal = optimize_weights_proximal_legacy
+
+
+# Mainly used to check if the group-size is divisible by numel()
+def is_divisible(val1: int, val2: int) -> bool:
+    return int(val2 * math.ceil(val1 / val2)) == val1
+
+
+# Converts hqq format W_dequant = (W_q - zero)*scale into affinequantized format: (W_q - mid_point)*scale_ao + zero_ao
+def convert_to_affinequantized_format(W_q, scale, zero, nbits, shape):
```
Comment (jerryzh168) [inline torchao/prototype/hqq/core.py:89]: nit: to be more precise this is int4 tensor_core_tiled layout format (for tinygemm), and this probably only works for `nbits == 4`?

File: torchao/dtypes/affine_quantized_tensor.py (hunk 1 above)
Comment (mobicham) [reply torchao/dtypes/affine_quantized_tensor.py:196]: Okay great, if you have some example I can follow I can do it

File: torchao/prototype/hqq/example.py (hunk 2 above)
Comment (mobicham) [reply torchao/prototype/hqq/example.py:52]: * mapping_type is needed to be `asymmetric` because it's only for asymmetric quantization. 
* `block` size is needed  
* `quant_min` / `quant_max` are needed as well since they define the range,
*  `zero_point_domain` should be float since we can't updated the zero-point with int, 
* `layout_type` doesn't depend on the method.
 
The only thing left is `preserve_zero`, I don't really know what it is doing exactly and it's not used in hqq. But when I was comparing with the default settings, `preserve_zero=True` which is the default was giving bad results, so I set it to False. 

So in short, all the args are needed except `preserve_zero` and `layout_type` since the default args for the rest do not match with hqq logic. 
There's a check, not sure if assert is enough here, let me know how you guys do it properly in torchao:
```Python
assert zero_point_domain == ZeroPointDomain.FLOAT and mapping_type == MappingType.ASYMMETRIC and quant_min==0, "Invalid input parameters for HQQ quantization."
```

File: torchao/quantization/hqq.py (hunk 3 above)
Comment (mobicham) [reply torchao/quantization/hqq.py:104]: Sure I can do it. The reason why I did it via a class because you can change the proximal solver settings globally by simply changing `HQQQuantizer.optimizer_weights = partial(...)`. If we stick to the default settings `optimize_weights_proximal_legacy` we don't need that. 
Will move it to a function in `quant_primitives` and we can decide later if we need another version with different settings.

File: torchao/quantization/hqq.py (hunk 4 above)
Comment (mobicham) [reply torchao/quantization/hqq.py:80]: Alright, this answers my previous question: https://github.com/pytorch/ao/pull/605#discussion_r1706611701

File: torchao/prototype/hqq/core.py (hunk 5 above)
Comment (mobicham) [reply torchao/prototype/hqq/core.py:89]: I am not sure, I was just following the code in the torchao repo and noticed that it uses the mid-point version. 
In the hqq quantize function, you can choose if you want this format or the default format (`W_q - zero)*scale` via the `raw_output` argument. What format is used for lowers bits in ao?

Comment (mobicham) [reply torchao/prototype/hqq/core.py:89]: So I updated the example to run with mutliple nbits and it looks like it's working fine with the mid-point format across all nbits: https://github.com/pytorch/ao/pull/605/commits/e0226544b6e099e30cef2762ebe8bfd4293e36a0

**Comment by mobicham** [comment]
hqq_ao
Llama3.1 8B Instruct lm-eval, the individual scores for each benchmark are a bit different (probably because of the mid-point thing), but the average score is the same

**Review by HDCharles — APPROVED** [review]
Can you modify https://github.com/pytorch/ao/blob/main/torchao/quantization/quant_api.py#L384

to have a new argument use_hqq default to False that applies the hqq quantization
so that it aligns with the existing API?

File: torchao/prototype/hqq/core.py (hunk 5 above)
Comment (jerryzh168) [reply torchao/prototype/hqq/core.py:89]: oh OK it should works for PlainLayoutType I think. it's an implementation detail in tinygemm though, where zero point domain is `ZeroPointDomain.FLOAT`

> What format is used for lowers bits in ao?

we haven't decided yet, but I think it will depend on benchmark results, for tinygemm the float zero_point domain (and preserve_zero=False) was chosen because it wants to use a fma instruction in cuda so it runs faster. and it's fine if we only care about linear quantization.

if we want to use quantization for conv, then we'd need to preserve zero: https://github.com/pytorch/ao/blob/5f35645bb75b67dd217c1cf17495ecb4a18b82fa/torchao/quantization/quant_primitives.py#L536 and set zero point domain to INT as well so that padding does not introduce loss

File: torchao/prototype/hqq/example.py (hunk 2 above)
Comment (jerryzh168) [reply torchao/prototype/hqq/example.py]: you when you set `zero_point_domain` to float, typically preserve_zero should be False as well.

File: torchao/quantization/hqq.py (hunk 3 above)
Comment (jerryzh168) [reply torchao/quantization/hqq.py]: yeah that's what I meant, thanks

File: torchao/dtypes/affine_quantized_tensor.py (hunk 1 above)
Comment (jerryzh168) [reply torchao/dtypes/affine_quantized_tensor.py:195]: one thing I can think of now is just to do `AffineQuantizedTensor.from_float_hqq` and use hqq quant primitive ops there

**Comment by jerryzh168** [comment]
> The problem has to do with the midpoint used in the dequantization logic by AffineQuantizedTensor which produces zero-point values with very low magnitude.

this is talking about accuracy? I can see you'd need this conversion for int4 tinygemm, but it's not required for other cases, if this is significant, maybe you can expose `raw_output` as a flag to the top level quant_api for hqq

File: torchao/dtypes/affine_quantized_tensor.py (hunk 1 above)
Comment (mobicham) [reply torchao/dtypes/affine_quantized_tensor.py:195]: Is this confirmed with @HDCharles or we keep the `use_hqq` flag?

File: torchao/prototype/hqq/core.py (hunk 5 above)
Comment (mobicham) [reply torchao/prototype/hqq/core.py:89]: I see, we can decide on this later The format would depend also on the cuda kernels for lower bits when they are available. Maybe we can follow the BitBlas approach, they have an argument `zeros_mode` which defines how the dequantization step is performed. They support different formats including `(Wq - zero)*scale` https://github.com/microsoft/BitBLAS/blob/main/docs/PythonAPI.md

**Comment by mobicham** [comment]
@jerryzh168 we can keep it this way for the moment. The other solution would be to modify the alternate minimization and update the quant parameters based ont he mid-point format instead of `(Wq - zero)*scale`. For the moment, the average lm-eval is about the same as hqq-lib, so I guess it's mostly fine. If there are some performance issues I can get back to it and try to change the steps to improve performance with the mid-point approach which is needed for using the tinygemm kernel.

File: torchao/dtypes/affine_quantized_tensor.py (hunk 1 above)
Comment (jerryzh168) [reply torchao/dtypes/affine_quantized_tensor.py:199]: yeah, this is fine for now I think

File: torchao/prototype/hqq/example.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+            )
+
+    linear_layer.weight = q_tensor_default
+    print("nbits", nbits, "| Default dequantization error", (W - q_tensor_default.dequantize()).abs().mean().item())
+    print("nbits", nbits, '| Default Dot product error', (y_ref - linear_layer(x.to(compute_dtype))).abs().mean().item())
+    # nbits 4 | Default dequantization error 0.001953125
+    # nbits 4 | Default Dot product error 0.005926903802901506
+
+
+    q_tensor_hqq = AffineQuantizedTensor.from_float(
```
Comment (jerryzh168) [inline torchao/prototype/hqq/example.py:60]: nit: I think you can use `to_affine_quantized` factory function here

File: torchao/prototype/hqq/example.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+linear_layer_default = torch.nn.Linear(in_features, out_features, bias=False, device=device)
+linear_layer_default.weight.data = W.clone()
+linear_layer_default = int4_weight_only_patch_fct(linear_layer_default)
+print("nbits", nbits, "| Default dequantization error", (W - linear_layer_default(torch.eye(W.shape[1], dtype=W.dtype, device=W.device)).T).abs().mean().item())
+print("nbits", nbits, '| Default Dot product error', (y_ref - linear_layer_default(x.to(compute_dtype))).abs().mean().item())
+# nbits 4 | Default dequantization error 0.000492095947265625
+# nbits 4 | Default Dot product error 0.0015244047390297055
+
+
+q_tensor_hqq = AffineQuantizedTensor.from_float(
```
Comment (jerryzh168) [inline torchao/prototype/hqq/example.py:102]: same here

File: torchao/quantization/quant_api.py [hunk 8]
```diff
@@ -409,7 +409,22 @@ def apply_int4_weight_only_quant(weight):
         zero_point_dtype = torch.bfloat16
         zero_point_domain = ZeroPointDomain.FLOAT
         layout_type = TensorCoreTiledLayoutType(inner_k_tiles=inner_k_tiles)
-        return to_affine_quantized(weight, mapping_type, block_size, target_dtype, quant_min, quant_max, eps, zero_point_dtype=zero_point_dtype, preserve_zero=preserve_zero, zero_point_domain=zero_point_domain, layout_type=layout_type)
+
+        if(use_hqq):
+            return  AffineQuantizedTensor.from_float(
```
Comment (jerryzh168) [inline torchao/quantization/quant_api.py:414]: and here

File: torchao/quantization/quant_primitives.py [hunk 9]
```diff
@@ -632,3 +633,172 @@ def _choose_qparams_affine(
     scale = torch.clamp(scale, min=eps)
 
     return scale.to(dtype=scale_dtype), zero_point.to(dtype=zero_point_dtype)
+
+
+#HQQ
+############################################################################
+# Shrinking operator (proximal operator for the lp norm)
+def shrink_lp_op(x: torch.Tensor, beta: float, lp_norm: float) -> torch.Tensor:
```
Comment (jerryzh168) [inline torchao/quantization/quant_primitives.py:641]: nit: please add `_` before ops that are not top level

File: torchao/quantization/quant_primitives.py [hunk 10]
```diff
@@ -632,3 +633,172 @@ def _choose_qparams_affine(
...
+    quant_min = 0
+    quant_max = 2**nbits - 1
+    mid_point = (quant_max + quant_min + 1) / 2
+    zero_ao = ((mid_point - zero.float()) * scale.float()).to(zero.dtype)
+    scale_ao = scale
+    W_q_ao = W_q.view(shape)
+    return W_q_ao, scale_ao, zero_ao
+
+#Main hqq quantizer function
+def quantize_affine_hqq(
```
Comment (jerryzh168) [inline torchao/quantization/quant_primitives.py:725]: nit: please add this to `__all__` as well in L21

**Review by jerryzh168 — APPROVED** [review]
looks good overall, just some nit comments, thanks @mobicham!

**Comment by mobicham** [comment]
Sorry for the delay, I was caught in other tasks. I will take a look at your comments now

File: torchao/quantization/quant_primitives.py (hunk 10 above)
Comment (mobicham) [reply torchao/quantization/quant_primitives.py:726]: https://github.com/pytorch/ao/pull/605/commits/9382ec12c703465710cea309fc4752af0d685f1e

File: torchao/quantization/quant_primitives.py (hunk 9 above)
Comment (mobicham) [reply torchao/quantization/quant_primitives.py]: https://github.com/pytorch/ao/pull/605/commits/41b2fb0745d20f44a03b5a3670dd26c60fd3dea1

File: torchao/quantization/quant_api.py (hunk 8 above)
Comment (mobicham) [reply torchao/quantization/quant_api.py]: https://github.com/pytorch/ao/pull/605/commits/3303d95522b8750d81b13e074ece3677890050b0

File: torchao/prototype/hqq/example.py (hunk 7 above)
Comment (mobicham) [reply torchao/prototype/hqq/example.py]: https://github.com/pytorch/ao/pull/605/commits/3303d95522b8750d81b13e074ece3677890050b0

File: torchao/prototype/hqq/example.py (hunk 6 above)
Comment (mobicham) [reply torchao/prototype/hqq/example.py]: https://github.com/pytorch/ao/pull/605/commits/3303d95522b8750d81b13e074ece3677890050b0

**Comment by jerryzh168** [comment]
@mobicham thanks, for tests right now we support both unittest.TestCase and pytest (e.g. https://github.com/pytorch/ao/blob/174e630af2be8cd18bc47c5e530765a82e97f45b/test/dtypes/test_uintx.py#L39), what you have in the file makes sense I think, since you have hardcoded errors for each dtype I feel it's fine to just separate separate them into different tests

**Comment by mobicham** [comment]
@jerryzh168 anything missing for the merge?

**Comment by msaroufim** [comment]
IIRC there was some test failure, just retriggered CI if all g ill merge

**Comment by mobicham** [comment]
@msaroufim the tests fail because of triton related to this. I was about to delete `torchao.prototype.hqq.core.py` since we move the hqq core stuff in quantization, but just saw this pr uses it : https://github.com/pytorch/ao/pull/679

```
  test/hqq/test_hqq_affine.py:3: in <module>
      from torchao.prototype.hqq.core import HQQQuantizer
  /opt/conda/envs/venv/lib/python3.9/site-packages/torchao/prototype/hqq/__init__.py:1: in <module>
      from .mixed_mm import triton_mixed_mm, pack_2xint4
  /opt/conda/envs/venv/lib/python3.9/site-packages/torchao/prototype/hqq/mixed_mm.py:2: in <module>
      import triton.language as tl
  E   ModuleNotFoundError: No module named 'triton'
```

**Comment by msaroufim** [comment]
I do see some legit looking errors for GPU https://github.com/pytorch/ao/actions/runs/10403091060/job/28816346879?pr=605#step:12:2570

And for CPU not sure I follow the connection to the version guard fix PR, basically we need to ensure HQQ Quantizer doesnt get imported by accident on CPU machines that way we can be sure triton is there

**Comment by mobicham** [comment]
It looks like it's the tensorcore test that is failing not the rest:
```Python
  test/hqq/test_hqq_affine.py::TestHQQ4bit::test_hqq_tensorcore_4bit FAILED
```
It was working fine just 2 days ago I think, let me clone and recheck on an instance

**Comment by mobicham** [comment]
> And for CPU not sure I follow the connection to the version guard fix PR, basically we need to ensure HQQ Quantizer doesnt get imported by accident on CPU machines that way we dont crash on the triton import

It's not supposed to be imported, I wanted to delete `torchao/prototype/hqq/core.py` because we move all the stuff there to `torchao/quantization/quant_primitives .py`. Will take a look at it

**Review by msaroufim — APPROVED** [review]

File: torchao/quantization/quant_api.py [hunk 11]
```diff
@@ -389,7 +389,7 @@ def int4_weight_only(group_size=128, layout_type=TensorCoreTiledLayoutType(inner
          size is more fine grained, choices are [256, 128, 64, 32]
         `layout_type`: layout type for quantized tensor, default is `TensorCoreTiledLayoutType(inner_k_tiles=8)`
     """
-    def apply_int4_weight_only_quant(weight):
+    def apply_int4_weight_only_quant(weight, use_hqq=False):
```
Comment (jerryzh168) [inline torchao/quantization/quant_api.py:392]: I just found that this flag is not used, so we don't really expose hqq to users right now, are you planning to create a new function for hqq? cc @mobicham

Comment (mobicham) [reply torchao/quantization/quant_api.py:392]: My understand was that @HDCharles suggested putting it there and later turning it on by default. 
It is exposed though via `to_affine_quantized` https://github.com/pytorch/ao/pull/605/files#diff-a9708dc28f15bb9cf665417e6c66601f9e8e2f1f672d1858603b74fa879a3357R62
Let me know if there's another way of exposing it

Comment (jerryzh168) [reply torchao/quantization/quant_api.py:392]: commented in https://github.com/pytorch/ao/pull/786#discussion_r1741153437
